src/lib.rs

```diff
@@ -78,6 +78,27 @@ pub trait OpenatDirExt {
     /// Remove all content at the target path, returns `true` if something existed there.
     fn remove_all<P: openat::AsPath>(&self, p: P) -> io::Result<bool>;
 
+    /// If `oldpath` exists, rename it to `newpath`. Otherwise do nothing.
+    ///
+    /// This returns `true` if the old path has been succesfully renamed, `false` otherwise.
+    fn local_rename_optional<P: AsRef<Path>, R: AsRef<Path>>(
+        &self,
+        oldpath: P,
+        newpath: R,
+    ) -> io::Result<bool>;
+
+    /// Try to rename `oldpath` to `newpath`, optionally preserving content already at `newpath`.
+    ///
+    /// If an entry exists at `oldpath` it will be renamed to the given `newpath`.
+    /// If the renaming is not possible because `newpath` is an already populated directory,
+    /// its content will be preserved and `oldpath` will be deleted instead (if empty).
+    /// This returns `true` if the old path has been succesfully renamed, `false` otherwise.
+    fn local_rename_conservative<P: AsRef<Path>, R: AsRef<Path>>(
```

Comment:
The intention of this library is definitely to have high level helpers, but this one feels too special case.   I have not seen an equivalent of this elsewhere. 

We don't have an equivalent of this in rpm-ostree do we?

Reply:
We do; this logic should be a straight port of rpm-ostree `rename_if_exists` (also linked in the commit message to ease side-by-side review): https://github.com/coreos/rpm-ostree/blob/v2021.4/src/libpriv/rpmostree-postprocess.cxx#L57

I do not disagree that this may not belong here, the logic is pretty arcane.

Reply:
When seeing this code again, I don't understand the "delete if empty" logic.  The commit message in
https://github.com/coreos/rpm-ostree/commit/45287828e5d1aa31fbebb6bbc176f6b37cc6f9da
doesn't touch on this.

I can't think of a situation in which this would occur or be helpful.  Can anyone?

So I think short term let's take one of two options:

- carry this helper in rpm-ostree
- try removing that logic entirely and if the tests pass, well I'm good to ship it

?

Reply:
I can't really say. The logic does make some sense ("move, but prefer keeping a populate target dir") but it really depends on the state and expectations of the underlying environment. According to @jlebon and https://github.com/coreos/rpm-ostree/pull/1589, this may (or may not) be needed for rpmdb handling.

> So I think short term let's take one of two options:
>  -  carry this helper in rpm-ostree
>  -  try removing that logic entirely and if the tests pass, well I'm good to ship it

I'll go for the former here.

Reply:
Right, I think the original reason that logic existed was for handling the `/var/lib/rpm` -> `/usr/share/rpm` transition. But nowadays we set things up so that librpm directly writes to the correct location already by setting `_dbpath`, so it should no longer be necessary.